conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
+	i := 0
+	for i < len(s.index) {
+		schema := s.index[i]
 		if schema.Pattern.MatchString(metric) {
-			return uint16(i), schema
+			// no interval passed,use the raw retentions.
+			// This is primarily used by the carbon input plugin.
+			if interval == 0 {
+				return uint16(i), *schema
+			}
+			// search through the retentions to find the first one where
+			// the metric interval is < SecondsPerPoint of the retention.
+			// The schema is then the previous retention.
+			for j, ret := range schema.Retentions {
+				if interval < ret.SecondsPerPoint {
+					// if there are no retentions with SecondsPerPoint >= interval
+					// then we need to use the first retention. Otherwise, the retention
+					// we want to use is the previous one.
+					if j > 0 {
+						j--
+					}
+					pos := i + j
+					return uint16(pos), *(s.index[pos])
+				}
+			}
+			// no retentions found with SecondsPerPoint > interval. So lets just use the retention
+			// with the largest secondsPerPoint.
+			pos := len(schema.Retentions) - 1 + i
+			return uint16(pos), *(s.index[pos])
 		}
+		// the next len(schema.Retentions) schemas in the index all have the
+		// same schema pattern, so we can skip over them.
+		i = i + len(schema.Retentions)
 	}
-	return uint16(len(s)), DefaultSchema
+
+	// as the DefaultSchema is in the schemas.index, this should typically never be reached.
+	// Though as the user can modify schemas.DefaultSchema we keep this for safety.
+	return uint16(len(s.index)), s.DefaultSchema
 }
 
 // Get returns the schema setting corresponding to the given index
 func (s Schemas) Get(i uint16) Schema {
-	if i+1 > uint16(len(s)) {
-		return DefaultSchema
+	if i+1 > uint16(len(s.index)) {
+		return s.DefaultSchema
+	}
+	return *(s.index[i])
+}
+
+// TTLs returns a slice of all TTL's seen amongst all archives of all schemas
+func (schemas Schemas) TTLs() []uint32 {
+	ttls := make(map[uint32]struct{})
```

Comment:
it could. this code predates the "permutation" stuff. There's no need for caching since this function rarely gets called and is not in any performance critical path.